File: core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala
/*
 * Copyright (2021) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.delta

import java.util.concurrent.TimeUnit

import scala.collection.mutable

import org.apache.spark.sql.delta.actions._
import org.apache.spark.sql.delta.metering.DeltaLogging
import org.apache.spark.sql.delta.util.FileNames

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.catalyst.expressions.{Expression, ExpressionSet, Or}
import org.apache.spark.sql.types.StructType

/**
 * A class representing different attributes of current transaction needed for conflict detection.
 *
 * @param readPredicates predicates by which files have been queried by the transaction
 * @param readFiles files that have been seen by the transaction
 * @param readWholeTable whether the whole table was read during the transaction
 * @param readAppIds appIds that have been seen by the transaction
 * @param metadata table metadata for the transaction
 * @param actions delta log actions that the transaction wants to commit
 * @param readSnapshot read [[Snapshot]] used for the transaction
 * @param commitInfo [[CommitInfo]] for the commit
 */
private[delta] case class CurrentTransactionInfo(
    val txnId: String,
    val readPredicates: Seq[DeltaTableReadPredicate],
    val readFiles: Set[AddFile],
    val readWholeTable: Boolean,
    val readAppIds: Set[String],
    val metadata: Metadata,
    val protocol: Protocol,
    val actions: Seq[Action],
    val readSnapshot: Snapshot,
    val commitInfo: Option[CommitInfo],
    val readRowIdHighWatermark: RowIdHighWaterMark,
    val domainMetadatas: Seq[DomainMetadata]) {

  /**
   * Final actions to commit - including the [[CommitInfo]] which should always come first so we can
   * extract it easily from a commit without having to parse an arbitrarily large file.
   *
   * TODO: We might want to cluster all non-file actions at the front, for similar reasons.
   */
  lazy val finalActionsToCommit: Seq[Action] = commitInfo ++: actions

  /** Whether this transaction wants to make any [[Metadata]] update */
  lazy val metadataChanged: Boolean = actions.exists {
    case _: Metadata => true
    case _ => false
  }


  /**
   * Partition schema corresponding to the read snapshot for this transaction.
   * NOTE: In conflict detection, we should be careful around whether we want to use the new schema
   * which this txn wants to update OR the old schema from the read snapshot.
   * e.g. the ConcurrentAppend check makes sure that no new files have been added concurrently
   * that this transaction should have read. So this should use the read snapshot partition schema
   * and not the new partition schema which this txn is introducing. Using the new schema can cause
   * issues.
   */
  val partitionSchemaAtReadTime: StructType = readSnapshot.metadata.partitionSchema

  def isConflict(winningTxn: SetTransaction): Boolean = readAppIds.contains(winningTxn.appId)
}

/**
 * Summary of the Winning commit against which we want to check the conflict
 * @param actions - delta log actions committed by the winning commit
 * @param commitVersion - winning commit version
 */
private[delta] class WinningCommitSummary(val actions: Seq[Action], val commitVersion: Long) {

  val metadataUpdates: Seq[Metadata] = actions.collect { case a: Metadata => a }
  val appLevelTransactions: Seq[SetTransaction] = actions.collect { case a: SetTransaction => a }
  val protocol: Seq[Protocol] = actions.collect { case a: Protocol => a }
  val commitInfo: Option[CommitInfo] = actions.collectFirst { case a: CommitInfo => a }.map(
    ci => ci.copy(version = Some(commitVersion)))
  val removedFiles: Seq[RemoveFile] = actions.collect { case a: RemoveFile => a }
  val addedFiles: Seq[AddFile] = actions.collect { case a: AddFile => a }
  val isBlindAppendOption: Option[Boolean] = commitInfo.flatMap(_.isBlindAppend)
  val blindAppendAddedFiles: Seq[AddFile] = if (isBlindAppendOption.getOrElse(false)) {
    addedFiles
  } else {
    Seq()
  }
  val changedDataAddedFiles: Seq[AddFile] = if (isBlindAppendOption.getOrElse(false)) {
    Seq()
  } else {
    addedFiles
  }
  val onlyAddFiles: Boolean = actions.collect { case f: FileAction => f }
    .forall(_.isInstanceOf[AddFile])

}

private[delta] class ConflictChecker(
    spark: SparkSession,
    initialCurrentTransactionInfo: CurrentTransactionInfo,
    winningCommitVersion: Long,
    isolationLevel: IsolationLevel) extends DeltaLogging {

  protected val startTimeMs = System.currentTimeMillis()
  protected val timingStats = mutable.HashMap[String, Long]()
  protected val deltaLog = initialCurrentTransactionInfo.readSnapshot.deltaLog

  protected var currentTransactionInfo: CurrentTransactionInfo = initialCurrentTransactionInfo

  protected val winningCommitSummary: WinningCommitSummary = createWinningCommitSummary()

  /**
   * This function checks conflict of the `initialCurrentTransactionInfo` against the
   * `winningCommitVersion` and returns an updated [[CurrentTransactionInfo]] that represents
   * the transaction as if it had started while reading the `winningCommitVersion`.
   */
  def checkConflicts(): CurrentTransactionInfo = {
    checkProtocolCompatibility()
    checkNoMetadataUpdates()
    checkForAddedFilesThatShouldHaveBeenReadByCurrentTxn()
    checkForDeletedFilesAgainstCurrentTxnReadFiles()
    checkForDeletedFilesAgainstCurrentTxnDeletedFiles()
    checkForUpdatedApplicationTransactionIdsThatCurrentTxnDependsOn()
    reassignOverlappingRowIds()
    checkIfDomainMetadataConflict()
    logMetrics()
    currentTransactionInfo
  }

  /**
   * Initializes [[WinningCommitSummary]] for the already committed
   * transaction (winning transaction).
   */
  protected def createWinningCommitSummary(): WinningCommitSummary = {
    recordTime("initialize-old-commit") {
      val winningCommitActions = deltaLog.store.read(
        FileNames.deltaFile(deltaLog.logPath, winningCommitVersion),
        deltaLog.newDeltaHadoopConf()
      ).map(Action.fromJson)
      new WinningCommitSummary(winningCommitActions, winningCommitVersion)
    }
  }

  /**
   * Asserts that the client is up to date with the protocol and is allowed to read and write
   * against the protocol set by the committed transaction.
   */
  protected def checkProtocolCompatibility(): Unit = {
    if (winningCommitSummary.protocol.nonEmpty) {
      winningCommitSummary.protocol.foreach { p =>
        deltaLog.protocolRead(p)
        deltaLog.protocolWrite(p)
      }
      if (currentTransactionInfo.actions.exists(_.isInstanceOf[Protocol])) {
        throw DeltaErrors.protocolChangedException(winningCommitSummary.commitInfo)
      }
    }
  }

  /**
   * Check if the committed transaction has changed metadata.
   */
  protected def checkNoMetadataUpdates(): Unit = {
    // Fail if the metadata is different than what the txn read.
    if (winningCommitSummary.metadataUpdates.nonEmpty) {
      throw DeltaErrors.metadataChangedException(winningCommitSummary.commitInfo)
    }
  }

  /**
   * Filters the [[files]] list with the partition predicates of the current transaction
   * and returns the first file that is matching.
   */
  protected def getFirstFileMatchingPartitionPredicates(files: Seq[AddFile]): Option[AddFile] = {
    import org.apache.spark.sql.delta.implicits._
    val filesDf = files.toDF(spark)

    // we need to canonicalize the partition predicates per each group of rewrites vs. nonRewrites
    val canonicalPredicates = currentTransactionInfo.readPredicates
      .partition(_.shouldRewriteFilter) match {
        case (rewrites, nonRewrites) =>
          val canonicalRewrites =
            ExpressionSet(rewrites.map(_.partitionPredicate)).map { e =>
              DeltaTableReadPredicate(partitionPredicates = Seq(e))
            }
          val canonicalNonRewrites =
            ExpressionSet(nonRewrites.map(_.partitionPredicate)).map { e =>
              DeltaTableReadPredicate(partitionPredicates = Seq(e), shouldRewriteFilter = false)
            }
          canonicalRewrites ++ canonicalNonRewrites
      }

    val filesMatchingPartitionPredicates = canonicalPredicates.iterator
      .flatMap { readPredicate =>
        DeltaLog.filterFileList(
          partitionSchema = currentTransactionInfo.partitionSchemaAtReadTime,
          files = filesDf,
          partitionFilters = readPredicate.partitionPredicates,
          shouldRewritePartitionFilters = readPredicate.shouldRewriteFilter
        ).as[AddFile].head(1).headOption
      }.take(1).toArray

    filesMatchingPartitionPredicates.headOption
  }

  /**
   * Check if the new files added by the already committed transactions should have been read by
   * the current transaction.
   */
  protected def checkForAddedFilesThatShouldHaveBeenReadByCurrentTxn(): Unit = {
    recordTime("checked-appends") {
      // Fail if new files have been added that the txn should have read.
      val addedFilesToCheckForConflicts = isolationLevel match {
        case WriteSerializable if !currentTransactionInfo.metadataChanged =>
          winningCommitSummary.changedDataAddedFiles // don't conflict with blind appends
        case Serializable | WriteSerializable =>
          winningCommitSummary.changedDataAddedFiles ++ winningCommitSummary.blindAppendAddedFiles
        case SnapshotIsolation =>
          Seq.empty
      }

      val fileMatchingPartitionReadPredicates =
        getFirstFileMatchingPartitionPredicates(addedFilesToCheckForConflicts)

      if (fileMatchingPartitionReadPredicates.nonEmpty) {
        val isWriteSerializable = isolationLevel == WriteSerializable

        val retryMsg = if (isWriteSerializable && winningCommitSummary.onlyAddFiles &&
          winningCommitSummary.isBlindAppendOption.isEmpty) {
          // The transaction was made by an older version which did not set `isBlindAppend` flag
          // So even if it looks like an append, we don't know for sure if it was a blind append
          // or not. So we suggest them to upgrade all there workloads to latest version.
          Some(
            "Upgrading all your concurrent writers to use the latest Delta Lake may " +
              "avoid this error. Please upgrade and then retry this operation again.")
        } else None
        throw DeltaErrors.concurrentAppendException(
          winningCommitSummary.commitInfo,
          getPrettyPartitionMessage(fileMatchingPartitionReadPredicates.get.partitionValues),
          retryMsg)
      }
    }
  }

  /**
   * Check if [[RemoveFile]] actions added by already committed transactions conflicts with files
   * read by the current transaction.
   */
  protected def checkForDeletedFilesAgainstCurrentTxnReadFiles(): Unit = {
    recordTime("checked-deletes") {
      // Fail if files have been deleted that the txn read.
      val readFilePaths = currentTransactionInfo.readFiles.map(
        f => f.path -> f.partitionValues).toMap
      val deleteReadOverlap = winningCommitSummary.removedFiles
        .find(r => readFilePaths.contains(r.path))
      if (deleteReadOverlap.nonEmpty) {
        val filePath = deleteReadOverlap.get.path
        val partition = getPrettyPartitionMessage(readFilePaths(filePath))
        throw DeltaErrors.concurrentDeleteReadException(
          winningCommitSummary.commitInfo, s"$filePath in $partition")
      }
      if (winningCommitSummary.removedFiles.nonEmpty && currentTransactionInfo.readWholeTable) {
        val filePath = winningCommitSummary.removedFiles.head.path
        throw DeltaErrors.concurrentDeleteReadException(
          winningCommitSummary.commitInfo, s"$filePath")
      }
    }
  }

  /**
   * Check if [[RemoveFile]] actions added by already committed transactions conflicts with
   * [[RemoveFile]] actions this transaction is trying to add.
   */
  protected def checkForDeletedFilesAgainstCurrentTxnDeletedFiles(): Unit = {
    recordTime("checked-2x-deletes") {
      // Fail if a file is deleted twice.
      val txnDeletes = currentTransactionInfo.actions
        .collect { case r: RemoveFile => r }
        .map(_.path).toSet
      val deleteOverlap = winningCommitSummary.removedFiles.map(_.path).toSet intersect txnDeletes
      if (deleteOverlap.nonEmpty) {
        throw DeltaErrors.concurrentDeleteDeleteException(
          winningCommitSummary.commitInfo, deleteOverlap.head)
      }
    }
  }

  /**
   * Checks [[DomainMetadata]] to capture whether the current transaction conflicts with the
   * winning transaction at any domain.
   *     1. Accept the current transaction if its set of metadata domains do not overlap with the
   *        winning transaction's set of metadata domains.
   *     2. Otherwise, fail the current transaction unless each conflicting domain is associated
   *        with a table feature that defines a domain-specific way of resolving the conflict.
   */
  private def checkIfDomainMetadataConflict(): Unit = {
    if (!DomainMetadataUtils.domainMetadataSupported(currentTransactionInfo.protocol)) {
      return
    }
    // Note [[currentTransactionInfo.domainMetadatas]] can't have duplicate domains otherwise
    // it would have failed [[validateDuplicateDomainMetadata]] at beginning of the transaction's
    // validation.
    val currentDomainMetadatas =
      currentTransactionInfo.domainMetadatas.map(action => action.domain -> action)
    val winningDomainMetadataMap =
      DomainMetadataUtils.extractDomainMetadatasMap(winningCommitSummary.actions)

    /**
     * Any new well-known domains that need custom conflict resolution need to add new cases in
     * below case match clause. E.g.
     * case ("delta.example.monotonicCounter", domain, Some(conflictDomain)) =>
     *   domain.copy(value = Math.max(domain.value, conflictingDomain.value))
     */
    val mergedDomainMetadatas = currentDomainMetadatas
      .map { case (name, domain) =>
        (name, domain, winningDomainMetadataMap.get(name)) match {
          // No-conflict case.
          case (_, domain, None) => domain
          case (name, _, Some(_)) =>
            // Any conflict not specifically handled by a previous case must fail the transaction.
            throw new io.delta.exceptions.ConcurrentTransactionException(
              s"A conflicting metadata domain $name is added.")
        }
      }

    // Update the [[DomainMetadata]] to the merged one.
    val updatedActions = currentTransactionInfo.actions.map {
      case domainMetadata: DomainMetadata =>
        // Note the find will always succeed, because [[mergedDomainMetadatas]] already contains
        // every domain from the transaction.
        mergedDomainMetadatas.find(_.domain == domainMetadata.domain).getOrElse(domainMetadata)
      case other => other
    }
    currentTransactionInfo = currentTransactionInfo.copy(
      domainMetadatas = mergedDomainMetadatas,
      actions = updatedActions)
  }

  /**
   * Checks if the winning transaction corresponds to some AppId on which current transaction
   * also depends.
   */
  protected def checkForUpdatedApplicationTransactionIdsThatCurrentTxnDependsOn(): Unit = {
    // Fail if the appIds seen by the current transaction has been updated by the winning
    // transaction i.e. the winning transaction have [[SetTransaction]] corresponding to
    // some appId on which current transaction depends on. Example - This can happen when
    // multiple instances of the same streaming query are running at the same time.
    if (winningCommitSummary.appLevelTransactions.exists(currentTransactionInfo.isConflict(_))) {
      throw DeltaErrors.concurrentTransactionException(winningCommitSummary.commitInfo)
    }
  }

  /**
   * Checks whether the Row IDs assigned by the current transaction overlap with the Row IDs
   * assigned by the winning transaction. I.e. this function checks whether both the winning and the
   * current transaction assigned new Row IDs. If this the case, then this check assigns new Row IDs
   * to the new files added by the current transaction so that they no longer overlap.
   */
  private def reassignOverlappingRowIds(): Unit = {
    // The current transaction should only assign Row Ids if they are supported.
    if (!RowId.isSupported(currentTransactionInfo.protocol)) return

    winningCommitSummary.actions.collectFirst {
      case RowIdHighWaterMark(winningHighWaterMark) =>
        // The winning transaction assigned conflicting Row IDs. Adjust the Row IDs assigned by the
        // current transaction as if it had read the result of the winning transaction.
        val readHighWaterMark = currentTransactionInfo.readRowIdHighWatermark.highWaterMark
        assert(winningHighWaterMark >= readHighWaterMark)
        val watermarkDiff = winningHighWaterMark - readHighWaterMark

        val actionsWithReassignedRowIds = currentTransactionInfo.actions.map {
          // We should only update the row IDs that were assigned by this transaction, and not the
          // row IDs that were assigned by an earlier transaction and merely copied over to a new
          // AddFile as part of this transaction. I.e., we should only update the base row IDs
          // that are larger than the read high watermark.
          case a: AddFile if a.baseRowId.exists(_ > readHighWaterMark) =>
            val newBaseRowId = a.baseRowId.map(_ + watermarkDiff)
            a.copy(baseRowId = newBaseRowId)

          case waterMark @ RowIdHighWaterMark(v) =>
            waterMark.copy(highWaterMark = v + watermarkDiff)

          case a => a
        }
      currentTransactionInfo = currentTransactionInfo.copy(
        actions = actionsWithReassignedRowIds,
        readRowIdHighWatermark = RowIdHighWaterMark(winningHighWaterMark))
    }
  }

  /** A helper function for pretty printing a specific partition directory. */
  protected def getPrettyPartitionMessage(partitionValues: Map[String, String]): String = {
    val partitionColumns = currentTransactionInfo.partitionSchemaAtReadTime
    if (partitionColumns.isEmpty) {
      "the root of the table"
    } else {
      val partition = partitionColumns.map { field =>
        s"${field.name}=${partitionValues(DeltaColumnMapping.getPhysicalName(field))}"
      }.mkString("[", ", ", "]")
      s"partition ${partition}"
    }
  }

  protected def recordTime[T](phase: String)(f: => T): T = {
    val startTimeNs = System.nanoTime()
    val ret = f
    val timeTakenMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTimeNs)
    timingStats += phase -> timeTakenMs
    ret
  }

  protected def logMetrics(): Unit = {
    val totalTimeTakenMs = System.currentTimeMillis() - startTimeMs
    val timingStr = timingStats.keys.toSeq.sorted.map(k => s"$k=${timingStats(k)}").mkString(",")
    logInfo(s"[$logPrefix] Timing stats against $winningCommitVersion " +
      s"[$timingStr, totalTimeTakenMs: $totalTimeTakenMs]")
  }

  protected lazy val logPrefix: String = {
    def truncate(uuid: String): String = uuid.split("-").head
    s"[tableId=${truncate(initialCurrentTransactionInfo.readSnapshot.metadata.id)}," +
      s"txnId=${truncate(initialCurrentTransactionInfo.txnId)}] "
  }
}

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ConflictChecker.java
/*
 * Copyright (2023) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.delta.kernel.internal.replay;

import static io.delta.kernel.internal.DeltaErrors.wrapEngineExceptionThrowsIO;
import static io.delta.kernel.internal.TableConfig.IN_COMMIT_TIMESTAMPS_ENABLED;
import static io.delta.kernel.internal.actions.SingleAction.*;
import static io.delta.kernel.internal.util.FileNames.deltaFile;
import static io.delta.kernel.internal.util.Preconditions.checkArgument;
import static io.delta.kernel.internal.util.Preconditions.checkState;
import static io.delta.kernel.utils.CloseableIterable.inMemoryIterable;
import static java.lang.String.format;

import io.delta.kernel.data.ColumnVector;
import io.delta.kernel.data.ColumnarBatch;
import io.delta.kernel.data.Row;
import io.delta.kernel.engine.Engine;
import io.delta.kernel.exceptions.ConcurrentWriteException;
import io.delta.kernel.internal.*;
import io.delta.kernel.internal.actions.CommitInfo;
import io.delta.kernel.internal.actions.DomainMetadata;
import io.delta.kernel.internal.actions.SetTransaction;
import io.delta.kernel.internal.util.DomainMetadataUtils;
import io.delta.kernel.internal.util.FileNames;
import io.delta.kernel.utils.CloseableIterable;
import io.delta.kernel.utils.CloseableIterator;
import io.delta.kernel.utils.FileStatus;
import java.io.*;
import java.util.*;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Class containing the conflict resolution logic when writing to a Delta table.
 *
 * <p>Currently, the support is to allow blind appends. Later on this can be extended to add support
 * for read-after-write scenarios.
 */
public class ConflictChecker {
  private static final int PROTOCOL_ORDINAL = CONFLICT_RESOLUTION_SCHEMA.indexOf("protocol");
  private static final int METADATA_ORDINAL = CONFLICT_RESOLUTION_SCHEMA.indexOf("metaData");
  private static final int TXN_ORDINAL = CONFLICT_RESOLUTION_SCHEMA.indexOf("txn");
  private static final int COMMITINFO_ORDINAL = CONFLICT_RESOLUTION_SCHEMA.indexOf("commitInfo");

  // Snapshot of the table read by the transaction that encountered the conflict
  // (a.k.a the losing transaction)
  private final SnapshotImpl snapshot;

  // Losing transaction
  private final TransactionImpl transaction;
  private final long attemptVersion;

  // Data actions from the losing transaction
  CloseableIterable<Row> dataActions;

  private ConflictChecker(
      SnapshotImpl snapshot,
      TransactionImpl transaction,
      long attemptVersion,
      CloseableIterable<Row> dataActions) {
    this.snapshot = snapshot;
    this.transaction = transaction;
    this.attemptVersion = attemptVersion;
    this.dataActions = dataActions;
  }

  /**
   * Resolve conflicts between the losing transaction and the winning transactions and return a
   * rebase state that the losing transaction needs to rebase against before attempting the commit.
   *
   * @param engine {@link Engine} instance to use
   * @param snapshot {@link SnapshotImpl} of the table when the losing transaction has started
   * @param transaction {@link TransactionImpl} that encountered the conflict (a.k.a the losing
   *     transaction)
   * @param dataActions {@link CloseableIterable} of {@link Row}s that represent the losing
   *     transaction's data actions
   * @return {@link TransactionRebaseState} that the losing transaction needs to rebase against
   * @throws ConcurrentWriteException if there are logical conflicts between the losing transaction
   *     and the winning transactions that cannot be resolved.
   */
  public static TransactionRebaseState resolveConflicts(
      Engine engine,
      SnapshotImpl snapshot,
      long attemptVersion,
      TransactionImpl transaction,
      CloseableIterable<Row> dataActions)
      throws ConcurrentWriteException {
    checkArgument(transaction.isBlindAppend(), "Current support is for blind appends only.");
    return new ConflictChecker(snapshot, transaction, attemptVersion, dataActions)
        .resolveConflicts(engine);
  }

  public TransactionRebaseState resolveConflicts(Engine engine) throws ConcurrentWriteException {
    List<FileStatus> winningCommits = getWinningCommitFiles(engine);
    AtomicReference<Optional<CommitInfo>> winningCommitInfoOpt =
        new AtomicReference<>(Optional.empty());

    // no winning commits. why did we get the transaction conflict?
    checkState(!winningCommits.isEmpty(), "No winning commits found.");

    FileStatus lastWinningTxn = winningCommits.get(winningCommits.size() - 1);
    long lastWinningVersion = FileNames.deltaVersion(lastWinningTxn.getPath());
    // Read the actions from the winning commits
    try (ActionsIterator actionsIterator =
        new ActionsIterator(engine, winningCommits, CONFLICT_RESOLUTION_SCHEMA, Optional.empty())) {

      actionsIterator.forEachRemaining(
          actionBatch -> {
            checkArgument(!actionBatch.isFromCheckpoint()); // no checkpoints should be read
            ColumnarBatch batch = actionBatch.getColumnarBatch();
            if (actionBatch.getVersion() == lastWinningVersion) {
              Optional<CommitInfo> commitInfo =
                  getCommitInfo(batch.getColumnVector(COMMITINFO_ORDINAL));
              winningCommitInfoOpt.set(commitInfo);
            }

            handleProtocol(batch.getColumnVector(PROTOCOL_ORDINAL));
            handleMetadata(batch.getColumnVector(METADATA_ORDINAL));
            handleTxn(batch.getColumnVector(TXN_ORDINAL));
            handleDomainMetadata(batch);
          });
    } catch (IOException ioe) {
      throw new UncheckedIOException("Error reading actions from winning commits.", ioe);
    }

    // if we get here, we have successfully rebased (i.e no logical conflicts)
    // against the winning transactions
    return new TransactionRebaseState(
        lastWinningVersion,
        getLastCommitTimestamp(
            engine, lastWinningVersion, lastWinningTxn, winningCommitInfoOpt.get()));
  }

  /**
   * Class containing the rebase state from winning transactions that the current transaction needs
   * to rebase against before attempting the commit.
   *
   * <p>Currently, the rebase state is just the latest winning version of the table. In future once
   * we start supporting read-after-write, domain metadata, row tracking, etc., we will have more
   * state to add. For example read-after-write will need to know the files deleted in the winning
   * transactions to make sure the same files are not deleted by the current (losing) transaction.
   */
  public static class TransactionRebaseState {
    private final long latestVersion;
    private final long latestCommitTimestamp;

    public TransactionRebaseState(long latestVersion, long latestCommitTimestamp) {
      this.latestVersion = latestVersion;
      this.latestCommitTimestamp = latestCommitTimestamp;
    }

    /**
     * Return the latest winning version of the table.
     *
     * @return latest winning version of the table.
     */
    public long getLatestVersion() {
      return latestVersion;
    }

    /**
     * Return the latest commit timestamp of the table. For ICT enabled tables, this is the ICT of
     * the latest winning transaction commit file. For non-ICT enabled tables, this is the
     * modification time of the latest winning transaction commit file.
     *
     * @return latest commit timestamp of the table.
     */
    public long getLatestCommitTimestamp() {
      return latestCommitTimestamp;
    }
  }

  /**
   * Any protocol changes between the losing transaction and the winning transactions are not
   * allowed. In future once we start supporting more table features on the write side, this can be
   * changed to handle safe protocol changes. For now the write support in Kernel is supported at a
   * very first version of the protocol.
   *
   * @param protocolVector protocol rows from the winning transactions
   */
  private void handleProtocol(ColumnVector protocolVector) {
    for (int rowId = 0; rowId < protocolVector.getSize(); rowId++) {
      if (!protocolVector.isNullAt(rowId)) {
        throw DeltaErrors.protocolChangedException(attemptVersion);
      }
    }
  }

  /**
   * Any metadata changes between the losing transaction and the winning transactions are not
   * allowed.
   *
   * @param metadataVector metadata rows from the winning transactions
   */
  private void handleMetadata(ColumnVector metadataVector) {
    for (int rowId = 0; rowId < metadataVector.getSize(); rowId++) {
      if (!metadataVector.isNullAt(rowId)) {
        throw DeltaErrors.metadataChangedException();
      }
    }
  }

  /**
   * Checks whether each of the current transaction's {@link DomainMetadata} conflicts with the
   * winning transaction at any domain.
   *
   * <ol>
   *   1. Accept the current transaction if its set of metadata domains does not overlap with the
   *   winning transaction's set of metadata domains.
   *   <p>2. Otherwise, fail the current transaction unless each conflicting domain is associated
   *   with a domain-specific way of resolving the conflict.
   * </ol>
   *
   * @param actionBatch action batch from the winning transactions
   */
  private void handleDomainMetadata(ColumnarBatch actionBatch) {
    // Extract the domain metadata map from the winning transaction.
    Map<String, DomainMetadata> winningDomainMetadataMap =
        DomainMetadataUtils.extractDomainMetadataMap(
            inMemoryIterable(actionBatch.getRows()), CONFLICT_RESOLUTION_SCHEMA);

    // Get the ordinal of the domainMetadata action from the full schema, which is used when writing
    // out the single action to the Delta Log
    final int domainMetadataOrdinal = FULL_SCHEMA.indexOf("domainMetadata");

    // Use try-with-resources to ensure that the CloseableIterable is closed after the loop
    try (CloseableIterable<Row> closeableDataActions = dataActions) {
      for (Row action : closeableDataActions) {
        // Skip non-domainMetadata actions
        if (action.isNullAt(domainMetadataOrdinal)) continue;

        // Extract DomainMetadata action that the losing transaction trying to commit
        DomainMetadata domainMetadata =
            DomainMetadata.fromRow(action.getStruct(domainMetadataOrdinal));

        // Try to resolve the conflict, if not possible, throw a ConcurrentTransaction exception
        resolveDomainMetadataConflict(domainMetadata, winningDomainMetadataMap);
      }
    } catch (IOException ioe) {
      throw new UncheckedIOException(ioe);
    }
  }

  private void resolveDomainMetadataConflict(
      DomainMetadata domainMetadataAttempt, Map<String, DomainMetadata> winningDomainMetadataMap) {

    String domain = domainMetadataAttempt.getDomain();
    DomainMetadata winningDomainMetadata = winningDomainMetadataMap.get(domain);
    if (winningDomainMetadata == null) {
      // No conflict
      return;
    } else {
      // Conflict - check if the conflict can be resolved

      // Currently, we don't have any domain-specific way of resolving the conflict.
      // Domain-specific ways of resolving the conflict can be added here (e.g. for Row Tracking)
      throw new ConcurrentWriteException(
          "A concurrent writer added a domainMetadata action for the same domain: "
              + domain
              + ". No domain-specific conflict resolution available for this domain."
              + "Attempted domainMetadata: "
              + domainMetadataAttempt
              + ". Winning domainMetadata: "
              + winningDomainMetadata);
    }
  }

  /**
   * Get the commit info from the winning transactions.
   *
   * @param commitInfoVector commit info rows from the winning transactions
   * @return the commit info
   */
  private Optional<CommitInfo> getCommitInfo(ColumnVector commitInfoVector) {
    for (int rowId = 0; rowId < commitInfoVector.getSize(); rowId++) {
      if (!commitInfoVector.isNullAt(rowId)) {
        return Optional.of(CommitInfo.fromColumnVector(commitInfoVector, rowId));
      }
    }
    return Optional.empty();
  }

  private void handleTxn(ColumnVector txnVector) {
    // Check if the losing transaction has any txn identifier. If it does, go through the
    // winning transactions and make sure that the losing transaction is valid from a
    // idempotent perspective.
    Optional<SetTransaction> losingTxnIdOpt = transaction.getSetTxnOpt();
    losingTxnIdOpt.ifPresent(
        losingTxnId -> {
          for (int rowId = 0; rowId < txnVector.getSize(); rowId++) {
            SetTransaction winningTxn = SetTransaction.fromColumnVector(txnVector, rowId);
            if (winningTxn != null
                && winningTxn.getAppId().equals(losingTxnId.getAppId())
                && winningTxn.getVersion() >= losingTxnId.getVersion()) {
              throw DeltaErrors.concurrentTransaction(
                  losingTxnId.getAppId(), losingTxnId.getVersion(), winningTxn.getVersion());
            }
          }
        });
  }

  private List<FileStatus> getWinningCommitFiles(Engine engine) {
    String firstWinningCommitFile =
        deltaFile(snapshot.getLogPath(), snapshot.getVersion(engine) + 1);

    try (CloseableIterator<FileStatus> files =
        wrapEngineExceptionThrowsIO(
            () -> engine.getFileSystemClient().listFrom(firstWinningCommitFile),
            "Listing from %s",
            firstWinningCommitFile)) {
      // Select all winning transaction commit files.
      List<FileStatus> winningCommitFiles = new ArrayList<>();
      while (files.hasNext()) {
        FileStatus file = files.next();
        if (FileNames.isCommitFile(file.getPath())) {
          winningCommitFiles.add(file);
        }
      }

      return ensureNoGapsInWinningCommits(winningCommitFiles);
    } catch (FileNotFoundException nfe) {
      // no winning commits. why did we get here?
      throw new IllegalStateException("No winning commits found.", nfe);
    } catch (IOException ioe) {
      throw new UncheckedIOException("Error listing files from " + firstWinningCommitFile, ioe);
    }
  }

  /**
   * Get the last commit timestamp of the table. For ICT enabled tables, this is the ICT of the
   * latest winning transaction commit file. For non-ICT enabled tables, this is the modification
   * time of the latest winning transaction commit file.
   *
   * @param engine {@link Engine} instance to use
   * @param lastWinningVersion last winning version of the table
   * @param lastWinningTxn the last winning transaction commit file
   * @param winningCommitInfoOpt winning commit info
   * @return last commit timestamp of the table
   */
  private long getLastCommitTimestamp(
      Engine engine,
      long lastWinningVersion,
      FileStatus lastWinningTxn,
      Optional<CommitInfo> winningCommitInfoOpt) {
    if (snapshot.getVersion(engine) == -1
        || !IN_COMMIT_TIMESTAMPS_ENABLED.fromMetadata(snapshot.getMetadata())) {
      return lastWinningTxn.getModificationTime();
    } else {
      return CommitInfo.getRequiredInCommitTimestamp(
          winningCommitInfoOpt, String.valueOf(lastWinningVersion), snapshot.getDataPath());
    }
  }

  private static List<FileStatus> ensureNoGapsInWinningCommits(List<FileStatus> winningCommits) {
    long lastVersion = -1;
    for (FileStatus commit : winningCommits) {
      long version = FileNames.deltaVersion(commit.getPath());
      checkState(
          lastVersion == -1 || version == lastVersion + 1,
          format(
              "Gaps in Delta log commit files. Expected version %d but got %d",
              (lastVersion + 1), version));
      lastVersion = version;
    }
    return winningCommits;
  }
}
